### PR TITLE
Initialization from rotation vector

### PR DESCRIPTION
**Pull request by adbugger**
Initial implementation and tests for initialization from rotation vector

### CODE REVIEW

File: scipy/spatial/transform/rotation.py [hunk 1]
```diff
@@ -195,3 +196,58 @@ def from_dcm(cls, dcm):
             return cls(quat[0], normalized=True)
         else:
             return cls(quat, normalized=True)
+
+    @classmethod
+    def from_rotvec(cls, rot_vecs):
+        """Initialize class from rotation vector.
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: rot_vecs -> rotvec for uniformity?

Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: rot_vecs -> rotvec for uniformity?

Apply everywhere.

File: scipy/spatial/transform/rotation.py [hunk 2]
```diff
@@ -195,3 +196,58 @@ def from_dcm(cls, dcm):
...
+            is_single = True
+
+        num_rotations = rot_vecs.shape[0]
+
+        norms = np.linalg.norm(rot_vecs, axis=1)[:, None]
+        small_angle = norms <= 1e-3
+
+        quat = np.empty((num_rotations, 4))
+
+        zero_norms = np.nonzero(norms == 0)[0]
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: No reason to consider this case separately from small_angle case.

File: scipy/spatial/transform/rotation.py [hunk 3]
```diff
@@ -195,3 +196,58 @@ def from_dcm(cls, dcm):
...
+        quat[zero_norms, :3] = np.array([0, 0, 0])
+
+        small_angle = np.nonzero(np.logical_and(
+            norms <= 1e-3, norms != 0))[0]
+        quat[small_angle, :3] = (0.5 - np.power(norms[small_angle], 2) /
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Don't know how much this is an improvement, but maybe introduce `scale = np.empty(n_rotations)`. Then `scale[small_angle] = ...` and `scale[large_angle] = ...` and finally `quat[:, :3] = scale[:, None] * rotvec`?

This way you also don't need to broadcast `norms` to 2d which is a slight improvement in clarity as well.

Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Why do you prefer to use `power` instead of `**`?

File: scipy/spatial/transform/tests/test_rotation.py [hunk 4]
```diff
@@ -200,3 +200,35 @@ def test_from_dcm_ortho_output():
...
+
+def test_from_2d_single_rotvec():
+    rotvec = [[1, 0, 0]]
+    expected_quat = np.array([[0.4794255, 0, 0, 0.8775826]])
+    result = Rotation.from_rotvec(rotvec)
+    assert_equal(result._single, False)
+    assert_array_almost_equal(result._quat, expected_quat)
+
+
+def test_from_generic_rotvec():
```
Comment (nmayorov) [inline scipy/spatial/transform/tests/test_rotation.py]: Add some tests with small (but non zero) angles as well. Think how to test it. 

My ideas is that `w` should be close to 1 (up to quadratic term) and `x, y, z` should be approximately `0.5 * rotvec`. Also check that the norm is very close to 1.

**Review by nmayorov — COMMENTED** [review]
Decent implementation, I suggested some possible improvements.

**Comment by larsoner** [comment]
@adbugger let me know when @nmayorov is happy and I'll take a look, too

File: scipy/spatial/transform/rotation.py (hunk 3 above)
Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: In this particular case, I read that the results of `np.power` and `**` can sometimes differ, not enough to be seen by `np.allclose` where tolerance is `1e-8` by default. Since I'm not sure how much of a concern this is, I'll use `**`.

File: scipy/spatial/transform/rotation.py [hunk 5]
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
...
+
+        small_angle = np.nonzero(np.logical_and(
+            norms <= 1e-3, norms != 0))[0]
+        scale[small_angle] = (0.5 - norms[small_angle] ** 2 / 48 +
+                              norms[small_angle] ** 4 / 3840)
+
+        large_angle = np.nonzero(norms > 1e-3)[0]
+        scale[large_angle] = (np.sin(norms[large_angle] / 2) /
+                              norms[large_angle])
+
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Let's create `quat` here.

File: scipy/spatial/transform/rotation.py [hunk 6]
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
...
+        # If a single vector is given, convert it to a 2D 1 x 3 matrix but
+        # set self._single to True so that we can return appropriate objects
+        # in the `as_...` methods
+        if rotvec.shape == (3,):
+            rotvec = rotvec[None, :]
+            is_single = True
+
+        num_rotations = rotvec.shape[0]
+
+        norms = np.linalg.norm(rotvec, axis=1)
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py:230]: I suggest to group `norms`, `small_angle` and `large_angle` in consecutive lines.

And then group `scale` and its element assignment.

File: scipy/spatial/transform/rotation.py [hunk 7]
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
...
+
+        quat = np.empty((num_rotations, 4))
+
+        scale = np.empty(num_rotations)
+
+        small_angle = np.nonzero(np.logical_and(
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Don't need to add `norm != 0` here.

File: scipy/spatial/transform/tests/test_rotation.py [hunk 8]
```diff
@@ -200,3 +201,57 @@ def test_from_dcm_ortho_output():
...
+        [0, 0, 0, 1]
+        ])
+    assert_array_almost_equal(
+            Rotation.from_rotvec(rotvec).as_quaternion(),
+            expected_quat)
+
+
+def test_from_rotvec_small_angle():
+    # Small theta
+    rotvec = np.array([1, 0.00001, 0.00001]) * 5e-6
```
Comment (nmayorov) [inline scipy/spatial/transform/tests/test_rotation.py]: The axis is not generic enough (along `x` practically) and I don't understand why you defined it this way. Could you test with a more interesting axis and perhaps somewhat lager angle?

Also we probably want to test on a stack of vectors where large and small angles are present.

File: scipy/spatial/transform/rotation.py [hunk 9]
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
...
+            norms <= 1e-3, norms != 0))[0]
+        scale[small_angle] = (0.5 - norms[small_angle] ** 2 / 48 +
+                              norms[small_angle] ** 4 / 3840)
+
+        large_angle = np.nonzero(norms > 1e-3)[0]
+        scale[large_angle] = (np.sin(norms[large_angle] / 2) /
+                              norms[large_angle])
+
+        quat[:, :3] = scale[:, None] * rotvec
+        quat[:, 3] = np.cos(norms / 2).flatten()
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Don't need to to flat it here.

**Review by nmayorov — COMMENTED** [review]
Some suggestions on how to reformat code to make it easier to follow.

Also try to improve tests for small angles.

File: scipy/spatial/transform/rotation.py [hunk 10]
```diff
@@ -195,3 +196,54 @@ def from_dcm(cls, dcm):
...
+
+        num_rotations = rotvec.shape[0]
+
+
+        scale = np.empty(num_rotations)
+
+        norms = np.linalg.norm(rotvec, axis=1)
+        small_angle = np.nonzero(norms <= 1e-3)[0]
+        large_angle = np.nonzero(norms > 1e-3)[0]
+
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Move `scale` here and I will be happy.

Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: I did exactly that in the latest commit :P

**Comment by nmayorov** [comment]
It looks good.

I have one idea about testing. My concern is that your tests contain some hard coded specific numbers, which is not necessary bad, but not ideal I guess.

We can test that the conversion is correct, if --- `w = cos(r / 2)` and `(x, y, z) is along r_` (can check with cross product) and `x**2 + y**2 + z**2 + w**2 = 1`. I don't suggest to rewrite your tests like this, unless you really like this idea. Maybe postpone if for later, when we have `to_rotvec`.

@larsoner also take a look.

@adbugger start working on `to_rotvec`.

File: scipy/spatial/transform/tests/test_rotation.py [hunk 11]
```diff
@@ -200,3 +201,71 @@ def test_from_dcm_ortho_output():
...
+    # cos(theta/2) ~~ 1 for small theta
+    assert_allclose(quat[0, 3], 1)
+    # sin(theta/2) / theta ~~ 0.5 for small theta
+    assert_allclose(quat[0, :3], rotvec[0] * 0.5)
+
+    assert_allclose(quat[1, 3], 0.9639685)
+    assert_allclose(
+            quat[1, :3],
+            np.array([
+                0.098796039321534656925304764958285802128222305393864012037,
```
Comment (nmayorov) [inline scipy/spatial/transform/tests/test_rotation.py]: Shorten these numbers somehow, 17 digits should be well enough.

File: scipy/spatial/transform/rotation.py [hunk 12]
```diff
@@ -195,3 +196,52 @@ def from_dcm(cls, dcm):
...
+        # set self._single to True so that we can return appropriate objects
+        # in the `as_...` methods
+        if rotvec.shape == (3,):
+            rotvec = rotvec[None, :]
+            is_single = True
+
+        num_rotations = rotvec.shape[0]
+
+        norms = np.linalg.norm(rotvec, axis=1)
+        small_angle = np.nonzero(norms <= 1e-3)[0]
```
Comment (larsoner) [inline scipy/spatial/transform/rotation.py]: no need for `np.nonzero` just use boolean masks directly. Then you can do:
```
small_angle = (norms <= 1e-3)
large_angle = ~small_angle
```
And the rest of the code should work

File: scipy/spatial/transform/rotation.py [hunk 13]
```diff
@@ -195,3 +196,52 @@ def from_dcm(cls, dcm):
...
+        large_angle = np.nonzero(norms > 1e-3)[0]
+
+        scale = np.empty(num_rotations)
+        scale[small_angle] = (0.5 - norms[small_angle] ** 2 / 48 +
```
Comment (larsoner) [inline scipy/spatial/transform/rotation.py]: A comment with a name or ref for the technique used here (taylor expansion of sin?) would be good. Think about the next dev who has to work with this code, they will probably not immediately know what this is doing

File: scipy/spatial/transform/tests/test_rotation.py [hunk 14]
```diff
@@ -200,3 +201,71 @@ def test_from_dcm_ortho_output():
...
+                0.09879603932153465,
+                0.14819405898230198,
+                0.19759207864306931
+                ]))
+
+    assert_equal(quat[2], np.array([0, 0, 0, 1]))
+
+
+def test_malformed_1d_from_rotvec():
+    with pytest.raises(ValueError):
```
Comment (larsoner) [inline scipy/spatial/transform/tests/test_rotation.py]: In general it's good practice to use `match='at least a few words from the error message'` here to ensure you catch the right error

**Review by larsoner — COMMENTED** [review]
Other than a few minor gripes, LGTM.

**Review by larsoner — APPROVED** [review]

**Comment by adbugger** [comment]
Should I merge this? The failures say internal error, I'm not sure what that means.

**Comment by adbugger** [comment]
Merging this due to review approval. I'm going to rebase `rotation_as_rotvec` off this, and open another pull request.
If required, I'll open another PR later to make fixes to this.
